### ISSUE TITLE
Uncontrolled input

### ISSUE THREAD

**Issue by Crysp** [preactjs/preact#1899]
Same logic with different libraries works different.

React - ✅
https://codesandbox.io/s/react-controlled-input-egxgu

Preact - ❌
https://codesandbox.io/s/preact-controlled-input-kib3p

**Comment by JoviDeCroock** [comment]
This is an inherent flaw to the inner workings for now, I'll see if I can find a solution to this problem but for now I can't see one yet.

Let me draw the situation,
1. for your first three keystrokes you pass in a different state
2. the fourth keystroke will return the same state meaning the component won't rerender
3. the HTMLelement however does the default behavior and appends a new keystroke

This can be prevented by doing the following, this is a workaround for now and I'll take a look if we can fix this without patching the event system:

```
return <input value={value} onInput={onChange} maxLength={3} />;
```

**Comment by dwelle** [comment]
@JoviDeCroock what does React do in this case? Binds `keydown` event and prevents it, whenever an input is controlled (there's `value={value}` on it)? Can't Preact do the same?

**Comment by JoviDeCroock** [comment]
React ships 50KB of event patching for these scenario's

**Comment by dwelle** [comment]
Right. Preventing `keydown` wouldn't actually help because `e.target.value` couldn't be used anymore and you'd need to set up workarounds around `e.which` and that opens a whole new can of worms and you end up with what React does. So no easy fix I can see either :/.

One possibility is to re-render even on same-state change when that state controls input value (not sure how easy would be to determine that).

**Comment by JoviDeCroock** [comment]
That's not possible without a flag or static code analysis (which is not possible with a runtime library)

**Comment by Crysp** [comment]
@JoviDeCroock this is simple example... what if i want to format received value?

`event.target.value.replace(/[^0-9]g/, '')`

**Comment by JoviDeCroock** [comment]
@Crysp nothing is stopping you from using maxLength for that too if I understand the problem you're trying to show well enough

**Comment by dwelle** [comment]
@JoviDeCroock he's right --- in his example it should prevent writing numbers which it fails to do so due to this bug. `maxLength` doesn't have a say in this.

Similarly, if you wanted to allow only letters, you wouldn't be able to do so:

https://codesandbox.io/s/preact-controlled-input-2-rvghz

**Comment by Crysp** [comment]
@JoviDeCroock as far as i understand, `set` doesn't use deep comparison for prev and next value. 

i want to write only numbers into state. if i input `abc` and then input `1` state, value of field resets to `1`. for first three characters `set` received empty string, compare default state and they are equal. but value of field still uncontrolled.

i think this not logical behaviour

**Comment by dwelle** [comment]
@Crysp it seems the only workaround is to leverage the behavior that `useState` does shallow comparison (`sameValueZero`), and use a non-primitive:

```js
const ControlledInput = () => {
  const [data, setData] = useState({ value: "" });

  const onChange = event =>
    setData({ value: event.target.value.replace(/[^a-z]/gi, "") });

  return <input value={data.value} onChange={onChange} />;
};
```

https://codesandbox.io/s/preact-controlled-input-2-rvghz

**Comment by JoviDeCroock** [comment]
@Crysp that's because you make your regexp reset the value to an empty string, this is nothing preact related.

- type "x"
- this makes event.target.value x
- type "1"
- this makes value x1 your regexp makes it ""

event.target.value.replace(/[^0-9]/g, "").slice(0, 3)

**Comment by dwelle** [comment]
@JoviDeCroock not quite, the regex simply removes forbidden characters which makes the `set()` result in no-op, and we're back at this bug ([react doesn't have it](https://codesandbox.io/s/react-controlled-input-68uv1)).

**Comment by Crysp** [comment]
@dwelle thx, i'll use your variant until bug will be fixed

**Comment by EirikFA** [comment]
I am having a similar issue with `value` simply being ignored. Setting it still allows any input.

React: https://codesandbox.io/s/gallant-frog-eeul9
Preact: https://codesandbox.io/s/nifty-brahmagupta-0z15g

**Comment by greasysock** [comment]
I am having issues with this as well, in the meantime I made this hook for creating controlled inputs:

```typescript
import * as React from 'react'

/** preact/compat controlled input fix */

/**
 * Preact/compat hook for fixing controlled inputs
 * @param value string value of the state to added to input
 * @param onChange callback method invoked when changing input
 */
export const useControlledInput = (value:string, onChange: (newValue: string) => void) => {
  const inputRef = React.useRef<HTMLInputElement>()

  React.useEffect(()=>{
    if(inputRef.current && value !== inputRef.current.value){
      inputRef.current.value = value || ''
    }
  },[value])

  React.useEffect(()=>{
    if(inputRef.current){
      // @ts-ignore - ts types are incorrect for onchange
      inputRef.current.oninput = (ev) => onChange(ev.target.value)
    }
  }, [onChange])

  return [
    /** Bind to ref for input */
    (el: HTMLInputElement|null) => {
      if(!inputRef.current && el){
        // @ts-ignore
        el.oninput = (ev) => onChange(ev.target.value)
        inputRef.current = el
        if(el.value !=== value) { el.value = value || '' }
      }
    }
  ]
}
```

example comp:

```javascript
import React, { useState } from 'react'
import { useControlledInput } from '../hooks'

const TestInput = () => {
  const [ controlledText, setControlledText ] = useState('')
  const [ controlledTextRef ] = useControlledInput( controlledText, (changes) => {
  // Do some validation
  setControlledText(changes)
} )
  return <input ref={controlledTextRef} />
}
```

**Comment by 18choi18** [comment]
I solved this issue by forcibly synchronizing the value of the input dom to the state.
`
<input value={x} onChange={evt => {
  const slicedValue = evt.target.value.slice(0, 3);
  setX(slicedValue);
  evt.target.value = slicedValue;
}} />
`

**Comment by ChenKS12138** [comment]
I also encountered the same problem and I solved it like this
```javascript
interface IInput {
  value: number | string;
  onInput: (e: any) => void;
  id: string;
  type: "text";
}

export default function Input(props: IInput) {
  const handleInput = useCallback(
    (event) => {
      const newValue = event.target.value;
      event.target.value = props.value;
      const newEvent = {
        ...event,
        target: { ...event.target, value: newValue },
      };
      props.onInput && props.onInput(newEvent);
    },
    [props.value]
  );

  return <input {...props} onInput={handleInput} />;
}

```

**Comment by fabiospampinato** [comment]
IMO this is a pretty significant issue, compatibility with React goes out of the window if stuff like this is broken. 

According to the [docs](https://preactjs.com/guide/v10/forms/#controlled--uncontrolled-components) controlled components actually work, IMO they should be updated to note that they don't actually work (at least for inputs it seems), especially considering how this issue is more than 2 years old, having the docs provide wrong information just makes people lose a bit of sanity.

**Comment by nullpaper** [comment]
As @fabiospampinato says, the problem here is that the docs are flat-out _misleading_. From the docs https://preactjs.com/guide/v10/forms/#controlled--uncontrolled-components:

> "The component doesn't manage the value itself there, but something else higher up in the component tree. ... Generally, you should try to use Controlled Components at all times."
> 
> ```jsx
> // Controlled, because Preact manages the input's value now
> <input value={someValue} onInput={myEventHandler} />;
> ```

Yeah, except if you actually try and and _control_ your input you get completely undocumented behaviour that works nothing like a "controlled input" is expected to. Nothing on that page even hints at the fact that Preact _doesn't actually control_ your input.

If I type `"1234"` into an example "controlled input" that replaces numbers, you get "1234" rickrolled:

```jsx
<input
  value={someValue}
  onInput={e => {
    const numbers = /[0-9]/g
    setSomeValue(e.replace(numbers, ''))
  }}
/>

// Lol this isn't controlled at all, nothing is replaced
```

Preact just **needs to explain this in the docs**.

**Comment by d-eisenga** [comment]
I'd like to know the status as well. Forcibly calling `e.target.value = x` on every input event moves the element's cursor to the end of the text, which is horrible UX, but there seems to be no other way to ensure that the input element doesn't become uncontrolled.

**Comment by scarf005** [comment]
also for signals, the workaround stops working as soon as computed values are used:

```tsx
const text = signal({})
const renderedText = computed(() => text.value.v)

const ControlledInput = () => {
	const onInput = e => {
		text.value = ({ v: e.currentTarget.value.replace(/[^\d]/, '') })
  }
	return <input value={renderedText} onInput={onInput} />
}
```

which feels like a serious footgun.

[repl link](https://preactjs.com/repl?code=aW1wb3J0IHugcmVuZGVyIH0gZnJvbSAicHJlYWN0IgppbXBvcnQge6BzaWduYWwsIGNvbXB1dGVkIH0gZnJvbSAiQHByZWFjdC9zaWduYWxzIgppbXBvcnQgeyB1c2VTdGF0ZSB9IGZyb20gInByZWFjdC9ob29rcyIKCmNvbnN0IHRleHQgPSBzaWduYWwoe30pCmNvbnN0IHJlbmRlcmVkVGV4dCA9IGNvbXB1dGVkKCgpID0%2BIHRleHQudmFsdWUudikKCmNvbnN0IENvbnRyb2xsZWRJbnB1dCA9ICgpID0%2BIHsKCWNvbnN0IG9uSW5wdXQgPSBlID0%2BIHsKCQl0ZXh0LnZhbHVlID0gKHsgdjogZS5jdXJyZW50VGFyZ2V0LnZhbHVlLnJlcGxhY2UoL1teXGRdLywgJycpIH0pCiAgfQoKCS8vIHRleHQudmFsdWUudgoJcmV0dXJuIDxpbnB1dCB2YWx1ZT17cmVuZGVyZWRUZXh0fSBvbklucHV0PXtvbklucHV0fSAvPgp9CgpyZW5kZXIoPENvbnRyb2xsZWRJbnB1dCAvPiwgZG9jdW1lbnQuZ2V0RWxlbWVudEJ5SWQoImFwcCIpKQo%3D)
